cmd/better-retester/retester.go

```diff
@@ -2,29 +2,95 @@ package main
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"io/ioutil"
+	"os"
 	"strings"
 	"time"
 
 	githubql "github.com/shurcooL/githubv4"
 	"github.com/sirupsen/logrus"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	"k8s.io/test-infra/prow/config"
 	"k8s.io/test-infra/prow/git/v2"
 	"k8s.io/test-infra/prow/github"
 	"k8s.io/test-infra/prow/tide"
+	"sigs.k8s.io/yaml"
 )
 
-type pullRequest struct {
-	prSha             string
-	baseSha           string
-	retestsForPrSha   int
-	retestsForBaseSha int
+type PullRequest struct {
+	PRSha              string      `json:"pr_sha,omitempty"`
+	BaseSha            string      `json:"base_sha,omitempty"`
+	RetestsForPrSha    int         `json:"retests_for_pr_sha,omitempty"`
+	RetestsForBaseSha  int         `json:"retests_for_base_sha,omitempty"`
+	LastConsideredTime metav1.Time `json:"last_considered_time,omitempty"`
 }
 
 type backoffCache struct {
-	cache map[string]*pullRequest
+	cache          map[string]*PullRequest
+	file           string
+	cacheRecordAge time.Duration
+	logger         *logrus.Entry
+}
+
+func (b *backoffCache) loadFromDisk() error {
+	if b.file == "" {
+		return nil
+	}
+	if _, err := os.Stat(b.file); errors.Is(err, os.ErrNotExist) {
+		b.logger.WithField("file", b.file).Info("cache file does not exit")
+		return nil
+	}
+	bytes, err := ioutil.ReadFile(b.file)
+	if err != nil {
+		return fmt.Errorf("failed to read file %s: %w", b.file, err)
+	}
+	cache := map[string]*PullRequest{}
+	if err := yaml.Unmarshal(bytes, &cache); err != nil {
+		return fmt.Errorf("failed to unmarshal: %w", err)
+	}
+	for key, pr := range cache {
+		if age := metav1.Now().Sub(pr.LastConsideredTime.Time); age > b.cacheRecordAge {
+			b.logger.WithField("key", key).WithField("LastConsideredTime", pr.LastConsideredTime.Time).
+				WithField("age", age).Info("deleting old record from cache")
+			delete(cache, key)
+		}
+	}
+	b.cache = cache
+	return nil
+}
+
+func (b *backoffCache) saveToDisk() (ret error) {
+	if b.file == "" {
+		return nil
+	}
+	bytes, err := yaml.Marshal(b.cache)
+	if err != nil {
+		return fmt.Errorf("failed to marshal: %w", err)
+	}
+	// write to a temp file and rename it to the cache file to ensure "atomic write":
+	// either it is complete or nothing
+	tmpFile, err := ioutil.TempFile("", "backoffCache")
+	if err != nil {
+		return fmt.Errorf("failed to create a temp file: %w", err)
+	}
+	tmp := tmpFile.Name()
+	defer func() {
+		if err := os.Remove(tmp); err != nil {
```

Comment:
Won't this error out by default when we succeed on `os.Rename` afterwards? After that the `tmp` path won't exist anymore.

Reply:
Updated.